proposals/0000-componentized-cabal.rst

```diff
@@ -0,0 +1,360 @@
+for the Cabal library (which is shared between cabal-install and Stack),
+and then a description of how cabal-install can take advantage of this
+new feature.  There is not a Stack portion to this proposal, but
+commentary from the Stack team would be appreciated.
+
+Proposed Change: Cabal
+----------------------
+
+**The status quo.** Currently, the Cabal library defines an command line
+API via the ``./Setup configure`` command. A conventional invocation of
+this command from, e.g., ``cabal-install``, might look like this::
+
+    ./Setup configure --ghc --prefix=$HOME/.cabal --user --ipid=package-name-1.0-abcdefg --dependency=array=array-0.5.1.0-abcdef1234567 --dependency=base=base-4.8.2.0-abcdef123456 --dependency=tasty=tasty-1.0-abcdef123456 --enable-tests --exact-configuration --disable-benchmarks
+
+There are a few important flags: ``--ipid`` specifies the *installed
+package identity* that the package will be built under: this is
+identifier is a unique identifier chosen by the package manager (for
+example, ``cabal new-build`` computes this identifier by hashing all of
+the possible inputs into the build, with the hope that the hash uniquely
+determines all of the build products of the package), and
+``--dependency=pkgname=ipid-1.0-abcdef`` specifies a specific dependency
+to be used during the build of all the components.
+``--exact-configuration`` requires ALL dependencies of all enabled
+components to be specified explicitly.  Notice that this particular
+command has ``--enable-tests`` passed (thus enabling the building of the
+test suite) and specifies which version of ``tasty`` (a test library)
+should be used. This command configures both the library and the test
+suite of this package *simultaneously*.
+
+Currently, the ``configure`` command takes no arguments.
+
+**The proposal.** We extend ``configure`` to take a single argument
+specifying which component is to be configured::
+
+    ./Setup configure test-suite
+
+It is invalid to pass multiple arguments. Under this mode, the configure
+command behaves differently in the following ways:
+
+1. Subsequent commands, e.g., build, *no longer* use an inplace package
+   database (conventionally in ``dist/package.conf.inplace``). It is
+   expected that all dependencies live in the database stack specified
+   by the user.
+
+2. The set of required ``--dependency`` flags for an
+   ``--exact-configuration`` is changed to precisely the ``build-depends``
+   of the component being enabled. Dependencies from non-enabled
+   components do not have to be specified (similar to how
+   ``--disable-tests`` works today). However, an *internal* dependency,
+   e.g., from a test suite to the library, must now be specified as
+   ``--dependency=packagename=pkgipid-1.0-abcdefg123``. In the presence of
+   convenience libraries, there may be multiple such extra "internal"
+   dependencies. These are disambiguated from ordinary dependencies by
```

Comment:
I'm ambivalent. It certainly wouldn't make the code any simpler: Cabal still has to do the legwork of interpreting package names as either internal or external dependencies.